### PR TITLE
Define a custom Scope Key for reducing memory allocations

### PR DESCRIPTION
**Pull request by sairamch04**
### Why ?

The current Scope Key is defined by creating a new string based on the metric tags and prefixes. This is inefficient as it ends up creating temporary strings for each metric emission. In our production service, the memory allocations due to scope key construction contribute to ~10% of the entire service allocations. This puts pressure on GC cycles, impacting the latencies of our service.

### What ?

Instead of defining the key based on a temporary string, we define it on already present objects. This will reduce unnecessary string allocations. 

We create a new POJO called `ScopeKey` which will be the key for HashMap. HashMap internally uses hashcode() and equal() methods to determine the key.

In short, We're changing the implementation of `ScopeImpl.keyForPrefixedStringMap` as part of this diff.

### Benchmark results

**Summary**

* Throughput increased from 1099 to 2306 Ops/Sec  (**~109% improvement**)
* Overall Memory allocation decreased from 1780 to 880 B/op (**~50% improvement**)
* Old Gen churn decreased from 0.24 to 0.115 B/op (**~50% improvement**)

**Before**

image

**After**

Screenshot 2023-02-08 at 12 37 49 AM

### CODE REVIEW

**Review by cdvr1993 — APPROVED** [review]
LGTM

File: core/src/test/java/com/uber/m3/tally/ScopeImplTest.java [hunk 1]
```diff
@@ -226,27 +226,34 @@ public void snapshot() {
 
         Snapshot snapshot = ((ScopeImpl) rootScope).snapshot();
 
-        Map<String, CounterSnapshot> counters = snapshot.counters();
+        Map<ScopeKey, CounterSnapshot> counters = snapshot.counters();
```
Comment (SokolAndrey) [inline core/src/test/java/com/uber/m3/tally/ScopeImplTest.java:229]: subsopes test is broken when i run individual test

File: core/src/main/java/com/uber/m3/tally/ScopeKey.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package com.uber.m3.tally;
+
```
Comment (SokolAndrey) [inline core/src/main/java/com/uber/m3/tally/ScopeKey.java:2]: afaik CI pipeline is broken, could you please run the tests and check style locally and fix the errors.
```
./gradlew checkstyleMain
```

**Review by SokolAndrey — COMMENTED** [review]
we also keep the file with other benchmarks here, since you're changing the core part of the library could you please re-run those benchmarks and see if you change affects those
```
./gradlew runJmhTests
```

**Comment by sairamch04** [comment]
@SokolAndrey : Thanks for the review. Ran all the benchmarks in tally-core and committed it. Please take a look again.

**Review by SokolAndrey — APPROVED** [review]
lgtm! looking forward to see it in production!
